src/lib.rs

```diff
@@ -1,53 +1,117 @@
+//! Inquirer – Fancy user interaction on the command line
+
+// Termion depends on nighty (for Keys), we need this for the unstable `std::io::CharsError` 
+#![feature(io)]
+
+#![deny(missing_docs,
+    missing_debug_implementations, missing_copy_implementations,
+    trivial_casts, trivial_numeric_casts,
+    unsafe_code,
+    unused_import_braces, unused_qualifications)]
+
 extern crate termion;
+#[macro_use] extern crate quick_error;
 
-use termion::{TermRead, TermWrite, IntoRawMode, color, Style, Key};
 use std::io::{Write, stdout, stdin};
+use std::fmt::Display;
+use termion::{TermRead, TermWrite, IntoRawMode, color, Style, Key};
+
+/// An option the user can choose
+///
+/// (Since the name "Option" is reserved for the well-known type representing
+/// nullability, we are calling this one "Choice".)
+pub trait Choice {
+    /// User visible text
+    type Text: Display;
+    /// Internal value representing this choice
+    type Value;
+
+    /// Get a reference to the text
+    fn text(&self) -> &Self::Text;
+
+    /// Get a reference to the value of this choice
+    fn value(&self) -> &Self::Value;
+}
+
+impl<'a> Choice for &'a str {
+    type Text = &'a str;
+    type Value = &'a str;
+
+    fn text(&self) -> &Self::Text {
+        self
+    }
+
+    fn value(&self) -> &Self::Value {
```

Comment:
Hm, maybe I mistook that for a simple example. What is a use case where you would like the user to select an option and then not get the text of the option but the index?

To restore the previous behaviour, you could make `list` not generic over `&[C]` but actually over an iterator yielding `C`s. This way, you could then call `.enumerate` when necessary to get the index.

Or tell the user to call `.enumerate().map(|index, item| (item, index))` themselves. (This way they are also seeing what values they'll be getting.)